debian/build_virtualenv

```diff
@@ -32,6 +32,8 @@ esac
 
 # Use --builtin-venv to use the better `venv` module from CPython 3.4+ rather
 # than the 2/3 compatible `virtualenv`.
+# Use the requirements exported from poetry so that we reproduce its locked
+# environment exactly. We generate requirements.txt in build_debian.sh.
```

Comment:
arg. This isn't really a thing we're meant to do.

The idea of debian packaging is that you can take a vanilla Debian VM, install the build dependencies, and expect `dpkg-buildpackage` to just work. Obviously, this adds dependencies on the environment.

A temporary workaround in the interests of landing this thing might be to *commit* requirements.txt (possibly inside the `debian` directory), but that's a thing we'll have to maintain manually. 

The correct solution is to run the `poetry export` here.

Reply:
> and expect dpkg-buildpackage to just work. 

So it does, kind of.

The impurity here is that our builds expect a `requirements.txt` to exist, which we're providing out of band (via `poetry export` in the builder container).

If that file is missing, like in the case of a raw `dpkg-buildpackage` invocation in a pristine Debian environment, then dh-virtualenv [silently skips it](https://github.com/spotify/dh-virtualenv/blob/6d5a12aae672fc7204ba4093e882cef871d58b14/dh_virtualenv/deployment.py#L201-L202) and then [eventually calls](https://github.com/spotify/dh-virtualenv/blob/6d5a12aae672fc7204ba4093e882cef871d58b14/dh_virtualenv/deployment.py#L270-L272) `pip install .` which has the effect of grabbing the latest compatible versions of all deps.

This obviously circumvents our attempts at dependency locking, but it does have the interesting property that "if a requirements.txt file is present, our `debian/rules` script will use it to constrain exactly which versions of dependencies are selected for the virtualenv."

@richvdh In ascending order of implementation difficulty, would you rather we:

1. Codify the above as intended behavior, updating comments / logging to be more explicit about what's happening.
2. Modify `debian/rules` (well, `debian/build_virtualenv`, really) to install Poetry from the Internet (probably into an ephemeral venv? Or using [pipx](https://pypa.github.io/pipx/)?), run `poetry export`, then nuke our Poetry installation
3. Modify all workflows / CI / etc., to ensure that we always have an up-to-date `poetry export > requirements.txt` checked into the repo whenever `poetry.lock` gets updated.
4. Manually package Poetry for all of our various Debian versions, much like we do dh-virtualenv (we can't use downstream poetry packages since the `export` command only works correctly for us in the 1.2 prereleases)

...or something else entirely?

Reply:
> If that file is missing, like in the case of a raw dpkg-buildpackage invocation in a pristine Debian environment, then dh-virtualenv [silently skips it](https://github.com/spotify/dh-virtualenv/blob/6d5a12aae672fc7204ba4093e882cef871d58b14/dh_virtualenv/deployment.py#L201-L202) and then [eventually calls](https://github.com/spotify/dh-virtualenv/blob/6d5a12aae672fc7204ba4093e882cef871d58b14/dh_virtualenv/deployment.py#L270-L272) pip install . which has the effect of grabbing the latest compatible versions of all deps.

hrm, that is interesting; I hadn't realised that; it does change my feelings here a bit.

I was expecting us to do option 2 (update `debian/build_virtualenv` to install and run Poetry, somehow), and that still feels like the cleanest option, but on reflection I could settle for option 1.

Reply:
To be honest, I was surprised at the silent success, but the siren song of "it's not a bug, it's a feature!"

(Though I do feel like it's a bug in dh-virtualenv which may eventually get fixed, biting us in the arse. For example, the dh-poetry project Shay found *never* worked as described, but appeared to due to that behavior)

Let's run with option 2... which will be janky in its own way, but dh-virtualenv itself is a bit of a rickety ship.

Reply:
Resolved in f0d72acebfd373b6428e8841898e90851a941192